parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java

```diff
@@ -519,6 +531,106 @@ public void appendFile(InputFile file) throws IOException {
     ParquetFileReader.open(file).appendTo(this);
   }
 
+  public int merge(List<InputFile> inputFiles, CodecFactory.BytesCompressor compressor, String createdBy, long maxBlockSize) throws IOException {
+    List<ParquetFileReader> readers = getReaders(inputFiles);
+    ByteBufferAllocator allocator = new HeapByteBufferAllocator();
+    ColumnReadStoreImpl columnReadStore = new ColumnReadStoreImpl(null, new DummyRecordConverter(schema).getRootConverter(), schema, createdBy);
+    this.start();
+    List<BlocksCombiner.SmallBlocksUnion> largeBlocks = BlocksCombiner.combineLargeBlocks(readers, maxBlockSize);
+    for (BlocksCombiner.SmallBlocksUnion smallBlocks : largeBlocks) {
+      for (int columnIndex = 0; columnIndex < schema.getColumns().size(); columnIndex++) {
+        ColumnDescriptor path = schema.getColumns().get(columnIndex);
+        ColumnChunkPageWriteStore store = new ColumnChunkPageWriteStore(compressor, schema, allocator);
+        ColumnWriteStoreV1 columnWriteStoreV1 = new ColumnWriteStoreV1(store, ParquetProperties.builder().build());
+        for (BlocksCombiner.SmallBlock smallBlock : smallBlocks.getBlocks()) {
+          ParquetFileReader parquetFileReader = smallBlock.getReader();
+          try {
+            Optional<PageReader> columnChunkPageReader = parquetFileReader.readColumnInBlock(smallBlock.getBlockIndex(), path);
+            ColumnWriter columnWriter = columnWriteStoreV1.getColumnWriter(path);
+            if (columnChunkPageReader.isPresent()) {
+              ColumnReader columnReader = columnReadStore.newMemColumnReader(path, columnChunkPageReader.get());
+              for (int i = 0; i < columnReader.getTotalValueCount(); i++) {
+                consumeTriplet(columnWriter, columnReader);
+              }
+            } else {
+              MessageType inputFileSchema = parquetFileReader.getFileMetaData().getSchema();
+              String[] parentPath = getExisingParentPath(path, inputFileSchema);
+              int def = parquetFileReader.getFileMetaData().getSchema().getMaxDefinitionLevel(parentPath);
+              int rep = parquetFileReader.getFileMetaData().getSchema().getMaxRepetitionLevel(parentPath);
+              for (int i = 0; i < parquetFileReader.getBlockMetaData(smallBlock.getBlockIndex()).getRowCount(); i++) {
+                columnWriter.writeNull(rep, def);
+              }
+            }
+          } catch (Exception e) {
+            LOG.error("File {} is not readable", parquetFileReader.getFile(), e);
+          }
+        }
+        if (columnIndex == 0) {
+          this.startBlock(smallBlocks.getRowCount());
+        }
+        columnWriteStoreV1.flush();
+        store.flushToFileWriter(path, this);
+      }
+      this.endBlock();
+    }
+    this.end(new HashMap<>());
+
+    BlocksCombiner.closeReaders(readers);
```

Comment:
Readers should be closed in a finally block to avoid unclosed readers in case of exception.

I was also wondering if it would be possible to keep only two files open at the same time: one which Parquet writes to, and one which is currently being processed. Since it seems sufficient for me to keep only these files open, and close the current one as soon as the processing of it is finished, and then open the next one, we won't keep files open longer than needed, and in addition we don't keep too many files open for no reason at the same time. I'm not sure if this makes sense, just a thought. If it is too complicated to implement, then I'm fine with the current approach too.